ehrql/query_engines/base_sql.py

```diff
@@ -85,6 +86,72 @@ def get_next_id(self):
         self.counter += 1
         return self.counter
 
+    def get_measure_queries(self, measures, results_query):
+        """
+        Return the SQL queries to fetch the results for a collection of measures
+        that share a denominator.
+
+        results_query is the result of calling get_queries on the dataset that
+        the measures will aggregate over.
+
+        The result is a query that object that is, or does the equivalent to,
+        GROUPING SETS for each measure's numerator and denominator aggregation
+        and group bys, with a GROUPING ID on all combined group bys. For each
+        measure, the value of GROUPING ID is an integer created by converting a
+        binary string of 0s and 1s for each group by column, where a 1 indicates
+        that the column is NOT a grouping column for that measure
+
+        e.g. we have 4 measures, and a total of 3 group by columns, [sex, region, ehnicity]
+        1) grouped by sex
+        2) grouped by region and ethnicity
+        3) grouped by sex, region and ethnicity
+
+        The grouping id for each of these would be:
+        1) 011 --> 3
+        2) 100 --> 4
+        3) 000 --> 0
+
+
+        """
+        measure_queries = []
+        all_group_by_cols = []
+        for i, measure in enumerate(measures.values()):
+            for group_by_col in measure.group_by:
+                group_col_query = results_query.c[group_by_col]
+                if group_col_query not in all_group_by_cols:
+                    all_group_by_cols.append(group_col_query)
+
+        for i, measure in enumerate(measures.values()):
+            # We need to return a column for each measure numerator and denominator in
+            # order to produce the same output columns as mssql's grouping sets
+            # We don't actually need to calculate the sums multiple times though
+            sum_overs = [sqlalchemy.null] * (len(measures) * 2)
+            sum_overs[(i * 2) : ((i * 2) + 2)] = [
+                sqlalchemy.func.sum(results_query.c[measure.sum_over[0]]).label(
+                    f"num_{i}"
+                ),
+                sqlalchemy.func.sum(results_query.c[measure.sum_over[1]]).label(
+                    f"den_{i}"
+                ),
+            ]
+
+            group_by_cols = [
+                results_query.c[group_by_col] for group_by_col in measure.group_by
+            ]
+            group_cols = [
+                gp if gp in group_by_cols else sqlalchemy.null
+                for gp in all_group_by_cols
+            ]
+            grouping_id = get_grouping_level_as_int(all_group_by_cols, group_by_cols)
+
+            measure_queries.append(
+                sqlalchemy.select(*sum_overs, *group_cols, grouping_id).group_by(
+                    *group_by_cols
```

Comment:
I don't think we actually need to include the group by columns anymore, we can infer them from the grouping id when we fetch the results. Not sure if that makes any difference though